### ISSUE TITLE
App fails to fix inconsistent database and repeatedly crashes

### ISSUE THREAD

**Issue by chreekat** [iSoron/uhabits#301]
### Background info

My device is a Nexus 5X, Android 7.1.2. However, this problem existed on a different phone, too.

I am on the beta tester program for Loop, and I have version 1.7 of the app. But again, this problem has persisted for a long time.

### Bug report

When I open the app, it shows the main screen and begins to populate the list of tasks. Within half a second, it closes, and Android tells me the app has unexpectedly closed. If I reopen the app, it starts successfully, and completely populates the list of tasks. From then on, everything is fine.

I open the app once or twice a day, and the bug happens nearly every time. However, if I reopen the app within a few minutes of the last time, it is fine.

**Comment by iSoron** [comment]
After the app crashes, can you please reopen it then send me a bug report from within the app (settings / generate bug report), so that I can have a better look?

**Comment by chreekat** [comment]
Will do!

**Comment by chreekat** [comment]
..and done.

**Comment by iSoron** [comment]
Thank you for the bug report, @chreekat. For the record, I have copied the relevant part of the bug report below and I am tagging this as confirmed.

```
06-07 11:20:56.501  3797  3820 W System.err: java.lang.RuntimeException: An error occurred while executing doInBackground()
06-07 11:20:56.501  3797  3820 W System.err: at android.os.AsyncTask$3.done(AsyncTask.java:325)
06-07 11:20:56.501  3797  3820 W System.err: at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
06-07 11:20:56.501  3797  3820 W System.err: at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
06-07 11:20:56.501  3797  3820 W System.err: at java.util.concurrent.FutureTask.run(FutureTask.java:242)
06-07 11:20:56.501  3797  3820 W System.err: at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
06-07 11:20:56.501  3797  3820 W System.err: at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
06-07 11:20:56.501  3797  3820 W System.err: at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
06-07 11:20:56.501  3797  3820 W System.err: at java.lang.Thread.run(Thread.java:761)
06-07 11:20:56.506  3797  3820 W System.err: Caused by: org.isoron.uhabits.models.sqlite.InconsistentDatabaseException: habit=***, 1 expected, 2 found
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.models.sqlite.SQLiteCheckmarkList.getByInterval(SQLiteCheckmarkList.java:122)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.models.CheckmarkList.getValues(CheckmarkList.java:135)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.models.ScoreList.forceRecompute(ScoreList.java:273)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.models.ScoreList.compute(ScoreList.java:213)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.models.ScoreList.getValue(ScoreList.java:86)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.models.ScoreList.getTodayValue(ScoreList.java:72)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.models.sqlite.SQLiteScoreList.getTodayValue(SQLiteScoreList.java:143)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.activities.habits.list.model.HabitCardListCache$RefreshTask.doInBackground(HabitCardListCache.java:321)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.tasks.AndroidTaskRunner$CustomAsyncTask.doInBackground(AndroidTaskRunner.java:108)
06-07 11:20:56.506  3797  3820 W System.err: at org.isoron.uhabits.tasks.AndroidTaskRunner$CustomAsyncTask.doInBackground(AndroidTaskRunner.java:86)
06-07 11:20:56.506  3797  3820 W System.err: at android.os.AsyncTask$2.call(AsyncTask.java:305)
06-07 11:20:56.506  3797  3820 W System.err: at java.util.concurrent.FutureTask.run(FutureTask.java:237)
06-07 11:20:56.506  3797  3820 W System.err: ... 4 more
```

**Comment by chreekat** [comment]
I saw the 'Repair database' option in settings, and clicked it... now I think I should wait to see if the problem happens again.

**Comment by chreekat** [comment]
I think repairing the database has made this problem go away!

I wish I had saved a copy of the broken database, so we could find the root cause of the problem I was having. Now, however, I think we should close this issue.

If it comes back, I can reopen.

**Comment by iSoron** [comment]
I will leave this open for a few days, in any case. If anyone else is facing this issue, please let us know.

**Comment by anowlcalledjosh** [comment]
I am also having this issue on Nexus 5X. One or more of the widgets on my home screen fail to display (I think the message is "Error displaying widget") almost every day until I launch the app twice -- the first time it exits immediately, the second time it loads like normal. I've tried running "Repair Database" a couple of times, but it doesn't seem to help.

**Comment by chreekat** [comment]
Yes, I also experience trouble with widgets.

Sorry I haven't had a chance to generate a database dump, iSoron :/

**Comment by iSoron** [comment]
I have released a new version (1.7.6) on Google Play that is supposed to fix this issue. It is currently only available to the beta testers, so you might need to [join our open beta to download it](https://play.google.com/apps/testing/org.isoron.uhabits). Please, give it a try and let me know if the problem is solved.

**Comment by chreekat** [comment]
Excellent. I'll let you know if it ever happens again.
